audio: first attempt at rewriting the channel converters.

icculus opened the pull request:
This is a not-quite-ready attempt to fix #5860. This adds 2.1 and 4.1 support, but also reworks the channel converters to hopefully work better: they no longer try to chain converters (mono -> stereo -> 5.1 -> 7.1, etc), which should be a little faster but more importantly remove little accumulated distortions, and also they (I think) have better conversions in general.

I based this conversion work on what FNA does, which is based on what XNA does, which is probably based on the MSDN link that was already at the top of our converter source code in a FIXME.  :)

Note that this changes the layout of 6.1 audio (which was added for that Kinect audio array) to match what the rest of SDL expects. We'll either need to add a special-case swizzle somewhere for that, or as the original kinect person suggested, just let kinect users deal with the channel ordering on their end.

There were a few conversions where I thought "that can't possibly be right" and made some adjustments, but now I'm second-guessing that; I'm going to go look at the mentioned MSDN link and see if I can learn some more before I finish this off.

Also, the handful of custom SIMD converters have not yet been updated.

And beyond forcing a 7.1 test WAV file to playback as stereo on my laptop, this hasn't gotten a lot of testing, and I don't currently have a good hardware setup to do so.

Still, my suspicion is that this is still better than the current state of things even with a few bugs to shake out.

flibitijibibo commented:
The repro for the matrix defaults, if anyone wants it, uses XAudio2:

```c
/* For all possible combinations... */
extern int srcChans, dstChans;
extern WAVEFORMATEX fmt; /* Ugh */
extern IXAudio2 *engine; /* Be sure to create a mastering voice too! */

IXAudio2SubmixVoice *submix;
IXAudio2SourceVoie *source;
XAUDIO2_VOICE_SENDS sends;
XAUDIO2_SEND_DESCRIPTOR sendDesc;
float matrix[9 * 9];
int i;

IXAudio2_CreateSubmixVoice(engine, &submix, dstChans, 48000, 0, 0, NULL, NULL);
sends.sendCount = 1;
sends.pSends = &sendDesc;
sendDesc.Flags = 0;
sendDesc.pOutputVoice = submix;
IXAudio2_CreateSourceVoice(engine, &source, &fmt, 0, 2.0f, NULL, &sends, NULL);
IXAudio2SourceVoice_GetOutputMatrix(source, submix, srcChans, dstChans, matrix);
IXAudio2SourceVoice_DestroyVoice(source);
IXAudio2SubmixVoice_DestroyVoice(submix);
for (i = 0; i < srcChans * dstChans; i += 1)
{
    printf("%.9f ", matrix[i]);
}
printf("\n");
```

slouken commented on src/audio/SDL_audiocvt.c:
```diff
@@ -63,61 +63,41 @@
+ * 6 channels (5.1) layout: FL+FR+FC+LFE+BL+BR
+ * 7 channels (6.1) layout: FL+FR+FC+LFE+BC+SL+SR
+ * 8 channels (7.1) layout: FL+FR+FC+LFE+BL+BR+SL+SR
+ */
```
For SDL 3, we should add a channel mask to the audio format structure, so applications can define which channels they're passing in, and can see what the hardware is expecting.

slouken commented on src/audio/SDL_audiocvt.c:
```diff
@@ -225,258 +205,979 @@ SDL_Convert51ToStereo_SSE(SDL_AudioCVT * cvt, SDL_AudioFormat format)
+    SDL_assert(format == AUDIO_F32SYS);
+
+    for (i = cvt->len_cvt / (sizeof (float) * 5); i; --i, src += 5) {
+        /* !!! FIXME: FNA/XNA mixes a little of the LFE into every channel...but this can't possibly be right, right...? I just drop the LFE and copy the channels. */
```
That actually makes sense to me. You don't want to lose all your bass when converting from X.1 to other formats.

icculus commented:
> Does it make sense to do this matrix-based instead of hard-coding a large number of converters?

That's exactly what FAudio does:

https://github.com/FNA-XNA/FAudio/blob/master/src/FAudio_internal_simd.c#L1281-L1301

But an enormous amount of this matrix is "multiply seven things by zero and something by one, and add them all together," so in the scalar path it's feels like a lot of wasted work.

flibitijibibo commented:
> Silly question... can we just crib the FNA code, leaving their license intact? Are we adding value by reimplementing this separately?

As FAudio's owner: Take whatever you want, we already do the same in the opposite direction (we reuse the bit depth converters since it's more direct than AudioCVT). It's all under zlib so it'll be a one-line attribution as far as we care!

icculus commented:
> Silly question... can we just crib the FNA code, leaving their license intact?

We can (it's also zlib), but generally my instinct is to avoid this sort of thing.

But I'll take a fast attempt at it and see what happens.

icculus commented:
Then again, this file in FAudio has pastes from SDL's audio data type converters and MojoAL's SIMD code, so maybe this is just how we do things.  :)

icculus commented:
I dunno, I'm skeptical this is a better path forward, at least for scalar fallbacks. The pull request, on x86-64 Linux, adds 28 kilobytes to the library in -O3 builds (20 in -Os), but it also handles every case...merging this FAudio code would likely _shrink_ the binary a little, I suspect...but it falls back to the generic version for many things, which may or may not be a problem in real life, but it's definitely going to be slower than the special-case code.

(And its SSE code is going to fall back to scalar in most cases, too...it can stay totally on the SIMD path for Quad and 7.1 input, do a partial sample frame in SSE for 5.1 and 6.1, and can't be used for mono or stereo input at all.)

The benefit to the FAudio code is that it's _smaller_, and the magic coefficients are all grouped into a single struct instead of scattered around the various functions, which isn't _clearer_ (one would never look at it and think "could this number be wrong?" because it's an opaque data block) but it does hide the complexity way better, and that's often worth something too, to be fair.

I would describe the FAudio approach as more _readable_, and that is good, but I think I'm going to clean up the current patch and carry on with that approach.

slouken commented:
Sounds good.

icculus commented:
(The Pull Request is failing checks because it has a debug flag turned on that calls fprintf, which isn't available. I'll fix that, too.)

icculus commented:
Okay, I ended up taking the exact matrix from FAudio, and built a little program to generate the converters. This means we won't have stupid human errors in the function variations, and maintenance is now a little program that calls printf instead of all the converters. Also, we can make changes to all these functions at once trivially.

The actual output is similar to what I had already built, so the same idea applies.

Here are some 5.1 and 7.1 test .wav files...

https://www.jensign.com/bdp95/7dot1voiced/7dot1voiced.zip

They just have a dude saying the channel number on each channel, in sequence, for testing purposes.

One notable thing right away: when running the 7.1 file through loopwave to my laptop's stereo internal speakers, it doesn't use SDL's converter at all, because PulseAudio accepts the device open at 7.1 and converts to stereo on our behalf.

And you can't hear the dude say "four" in the playback at all, and "five" is pretty quiet.

But you can hear it really well with SDL's new channel converter!

Simple program to force whatever .WAV file to convert to stereo with SDL's converter and play it:

```c
#include "SDL.h"

int main(int argc, char **argv)
{
    const char *filename = argv[1];
    SDL_AudioSpec devspec;
    SDL_AudioSpec wavspec;
    Uint8 *sound = NULL;
    Uint32 soundlen = 0;
    SDL_AudioStream *stream = NULL;

    if (!filename) {
        return 0;
    }

    SDL_Init(SDL_INIT_AUDIO);

    SDL_zero(wavspec);
    if (SDL_LoadWAV(filename, &wavspec, &sound, &soundlen) == NULL) {
        SDL_Log("Failed to load %s: %s", filename, SDL_GetError());
        SDL_Quit();
        return 1;
    }

    SDL_memcpy(&devspec, &wavspec, sizeof (devspec));
    devspec.channels = 2;
    devspec.samples = 4096;
    SDL_OpenAudio(&devspec, NULL);

    stream = SDL_NewAudioStream(wavspec.format, wavspec.channels, wavspec.freq,
                                devspec.format, devspec.channels, devspec.freq);

    SDL_AudioStreamPut(stream, sound, soundlen);
    SDL_AudioStreamFlush(stream);

    SDL_FreeWAV(sound);

    soundlen = SDL_AudioStreamAvailable(stream);
    sound = (Uint8 *) SDL_malloc(soundlen);
    SDL_AudioStreamGet(stream, sound, soundlen);
    SDL_QueueAudio(1, sound, soundlen);
    SDL_free(sound);
    SDL_FreeAudioStream(stream);
    SDL_PauseAudio(0);

    while (SDL_GetQueuedAudioSize(1) > 0) {
        SDL_Event e;
        while (SDL_PollEvent(&e)) { if (e.type == SDL_QUIT) { SDL_ClearQueuedAudio(1); } }
        SDL_Delay(100);
    }

    SDL_CloseAudio();
    SDL_Quit();

    return 0;
}
```

flibitijibibo commented:
Everything seems right to me, but I'm used to these coefficients so of course I'd say that :P

The one block that may need special treatment in SDL's case is its handing of stereo upmixing. For whatever reason they don't try to blend stereo (and stereo specifically...?) into any added channels, so we may want to come up with coefficients that make stereo upmixing use all the channels.

icculus commented:
Dropping the AVX 5.1->stereo converter; I rewrote it to use the coefficients, but it's permute hell to keep moving the data around to get six values organized in a vector of eight, horizontal-sum the vectors and then move it around again to get the final float in the right position of the output vector...it isn't any faster than the scalar version, if we want correct output.

I'm not spending the hours to rewrite the SSE and NEON equivalents to find out they're going to have the same problem, so they are going, too.

The SSE3 stereo-to-mono is still correct and can stay.

icculus commented:
Okay, added a mono-to-stereo SSE path, because it's a simple and obvious slam dunk for a likely codepath, fixed a few other things, and now this is done! Going to merge this when the checks finish, so speak now if you have any objections!

flibitijibibo reviewed (approved):
All lgtm. The only possible objection I could make is distributing stereo to all channels (except LFE) when upmixing, but the argument could be made that the stock conversion is more appropriate. I don't know what the ideal coefficients for that would be though, so at most this is for future patch.

icculus commented:
> The only possible objection I could make is distributing stereo to all channels

Oh, I forgot to address this before, sorry.

So this was one of the things that came up in discussing this with Valve: distributing stereo output around surround layouts turns out to be unpleasant for various reasons. I haven't really explored it, but I suspect there's a reason the XAudio2 matrix doesn't try it, either.
